Fix streaming eslint configuration

ThisIsMissEm opened the pull request:
This fixes some issues I noticed whilst working on #27828, whilst these didn't cause type-checking errors via linting, the issues were definitely present within rich IDEs such as VS Code.

This also gives us configuration completely independent of the frontend, allowing us to migrate streaming to typescript at a later date to the frontend or migrate streaming to ESM before the frontend.

ThisIsMissEm committed:
Fix eslint/tsconfig for streaming

ThisIsMissEm committed:
Streaming: allow unused labels via tsconfig, as these are common with node.js event handlers

renchap commented on streaming/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+    "target": "esnext",
+    "module": "CommonJS",
+    "moduleResolution": "node",
+    "allowUnusedLabels": true,
```
Labels are not commonly used in Javascript, are we sure we want this enabled?

ThisIsMissEm commented on .eslintrc.js:
```diff
@@ -1,3 +1,5 @@
+// @ts-check
+/** @type {import('eslint-define-config').ESLintConfig} */
```
We could do that too I guess? I just figured a type declaration was the softer touch approach?

renchap replied:
I think the function simply returns the argument, but it specifies its type so it is checked for validity.

This is the recommended way in `eslint-define-config`, so I wondered if there was a reason not to use it.

ThisIsMissEm replied:
@renchap should we save this change for later? I'm not sure if changing to defineConfig would cause conflicts with anything else in flight

ThisIsMissEm replied:
Okay, have made the change, but it had a cost.

ThisIsMissEm commented on tsconfig.json:
```diff
@@ -19,5 +19,6 @@
     "app/javascript/types"
-  ]
+  ],
+  "exclude": ["streaming/"]
```
Yeah, because for some reason this still applied to streaming, I'm not sure why though.

ThisIsMissEm replied:
Okay, I've removed this for now, hopefully it's fine!

renchap commented:
Also, as this project is not using Typescript but we like having types, should we have a [`jsconfig.json`](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) rather than a `tsconfig.json`? 

This would avoid having to specify `// @ts-check` for every file while still having Typescript process those.

ThisIsMissEm commented:
> Also, as this project is not using Typescript but we like having types, should we have a [`jsconfig.json`](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) rather than a `tsconfig.json`?
> 
> This would avoid having to specify `// @ts-check` for every file while still having Typescript process those.

I think I'd rather stick with tsconfig, the `// @ts-check` is a minor nuisance, and the goal is to eventually get to typescript, but it's a breaking change and kinda complex to do right now with the current way the code is. Consistent tooling is probably best.

ThisIsMissEm committed:
Fix issues noted in code review

ThisIsMissEm commented:
Have switched to [`noUnusedParameters: false`](https://www.typescriptlang.org/tsconfig#noUnusedParameters) instead, and removed the exclude which was needed when I did the change but no longer seems to be needed:

ThisIsMissEm committed:
Eslint: Use defineConfig

ThisIsMissEm commented on streaming/.eslintrc.js:
```diff
@@ -0,0 +1,29 @@
+    'import/no-extraneous-dependencies': [
+      'error',
+      {
+        devDependencies: [
+          'streaming/.eslintrc.js',
+        ],
```
Due to this, we need eslint-define-config in both streaming and the root package.json, which is annoying.

ThisIsMissEm commented on streaming/.eslintrc.js:
```diff
@@ -0,0 +1,29 @@
+        optionalDependencies: false,
+        peerDependencies: false,
+        includeTypes: true,
+        packageDir: __dirname,
```
It's more likely due to packageDir, but it's not possible to set this for just streaming but include devDependencies from the root package.

renchap commented on streaming/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+    "target": "esnext",
+    "module": "CommonJS",
+    "moduleResolution": "node",
+    "noUnusedParameters": false,
```
I would prefer to prefix unused parameters with `_` rather than setting this. IIRC Typescript should not complain about them, as the `_` prefix explicitly marks them as non-used

ThisIsMissEm replied:
Yeah, this is more, I didn't want to change all the method signatures at once, so allowing this for now is the better route.

renchap replied:
But are there functions with unused parameters at the moment?

I removed this and could not find any

renchap commented on .eslintrc.js:
```diff
@@ -193,6 +196,8 @@ module.exports = {
       {
         devDependencies: [
+          '.eslintrc.js',
+          'streaming/.eslintrc.js',
```
I am unsure why we need to include files from `streaming/` in the root config, it seems weird to me.

I will see if I find another way

ThisIsMissEm replied:
```suggestion
```

renchap commented:
I tried these changes and everything seems to work for me?

https://github.com/mastodon/mastodon/commit/af3254d2c497ac357553dd869fb4cc2900d56c9d

I also added support for `yarn typecheck` in `streaming/`, it does not pass at the moment but it shows the issues.

ThisIsMissEm commented on .eslintrc.js:
```diff
@@ -299,6 +304,8 @@ module.exports = {
     {
       files: [
+        '.eslintrc.js',
+        'streaming/.eslintrc.js',
```
```suggestion
```

ThisIsMissEm commented on streaming/.eslintrc.js:
```diff
@@ -0,0 +1,29 @@
+module.exports = defineConfig({
+  extends: ['../.eslintrc.js'],
+  parserOptions: {
+    project: true,
```
```suggestion
    project: true,
    env: {
      browser: false,
    },
```

ThisIsMissEm committed:
Apply suggestions from code review

These changes are taken from https://github.com/mastodon/mastodon/commit/af3254d2c497ac357553dd869fb4cc2900d56c9d

ThisIsMissEm commented on streaming/.eslintrc.js:
```diff
@@ -0,0 +1,32 @@
+  extends: ['../.eslintrc.js'],
+  parserOptions: {
+    project: true,
+    env: {
+      browser: false,
+    },
```
```suggestion
```

ThisIsMissEm commented on streaming/.eslintrc.js:
```diff
@@ -0,0 +1,32 @@
+
+module.exports = defineConfig({
+  extends: ['../.eslintrc.js'],
+  parserOptions: {
```
```suggestion
  env: {
    browser: false,
  },
  parserOptions: {
```

ThisIsMissEm committed:
Apply suggestions from code review

I'm obviously half asleep

ThisIsMissEm committed:
Streaming: Add check:types command

ThisIsMissEm commented on streaming/package.json:
```diff
@@ -12,7 +12,8 @@
   "scripts": {
-    "start": "node ./index.js"
+    "start": "node ./index.js",
+    "check:types": "tsc --noEmit"
```
@renchap I am dubious of the value this brings right now, so let's not make it gating.

renchap replied:
I never asked to have it ran anywhere, my goal was to have a way for devs to run it and try to get the number of errors down :)

renchap reviewed (approved):
Looks good to me.

I would prefer to not have `noUnusedParameters: true`, but I will make another PR to change the 5 places where this is triggered if you do not want it in this PR.

ThisIsMissEm commented:
> Looks good to me.
> 
> I would prefer to not have `noUnusedParameters: true`, but I will make another PR to change the 5 places where this is triggered if you do not want it in this PR.

I agree we need to fix this, but it is really a low priority, and conflicts with in-flight work, so let's hold off for a little bit (maybe an issue to track it?)
